### PR TITLE
ENH: Expose authentication parameters for DICOMweb networking

### PR DESCRIPTION
**Pull request by tbirdso**
These changes aim to expose authentication parameters in Python utilities
to aid in DICOM networking while avoiding implementation-specific
changes to the Slicer user interface. Generic methods such as a
user/password combination, OAuth token, or other application-specific
token can now (theoretically) be used in DICOMweb utility methods.

- A username/password combination can be persisted as a global setting
  with the `DICOMWEBUSER` and `DICOMWEBPWD` keys.
- A generic `requests.auth.AuthBase` object can be passed into Slicer
  utility methods for use in retrieving and stowing DICOM objects.
- DICOMSender is refactored for clarity. The DICOMSender public interface is
  limited to initialization and the `send` method. A DICOMweb
  authentication parameter is exposed, otherwise no functional changes
  are made.
- A method for verifying a Python module runtime requirement is
  generalized and moved from `DICOMSender` to `slicer.util`.

Manual testing results:
- DIMSE: ran `storescp` locally, exported sample data from the Slicer DICOM database to localhost and confirmed files populated in the storescp working directory
- DICOMweb: Ran a local [Orthanc](https://www.orthanc-server.com/download-windows.php) instance, confirmed that export with DICOMSender fails with missing/incorrect global authentication information and succeeds with correct auth information.

### CODE REVIEW

File: Modules/Scripted/DICOMLib/DICOMUtils.py
Comment (jcfr) [inline Modules/Scripted/DICOMLib/DICOMUtils.py:928]:
```suggestion
```

File: Modules/Scripted/DICOMLib/DICOMUtils.py
Comment (jcfr) [inline Modules/Scripted/DICOMLib/DICOMUtils.py:916]:
```suggestion
```

File: Modules/Scripted/DICOMLib/DICOMUtils.py
Comment (jcfr) [inline Modules/Scripted/DICOMLib/DICOMUtils.py:918]:
```suggestion
GLOBAL_DICOMWEB_USER_KEY = "GLOBAL_DICOMWEB_USER_KEY"
GLOBAL_DICOMWEB_PASSWORD_KEY = "GLOBAL_DICOMWEB_PASSWORD_KEY"
```

**Review by jcfr — COMMENTED** [review]
Thanks for the contribution.

Waiting we can further test, here are a few minor comments

File: Base/Python/slicer/util.py
Comment (tbirdso) [reply Base/Python/slicer/util.py:3802]: I agree with the suggested flake8 configurations, but I'm not sure I understand the suggested change to remove the `None` return type annotation? IMO that type hint is preferable to indicate that the `None` return value is intended, such as in this case where a validation function will throw an error, produce a side effect (restart the app), or otherwise do nothing. Removing the return type hint indicates that the return type is unknown, which is not correct.

Note that `None` return type is used throughout Python examples: https://peps.python.org/pep-0484/#type-aliases

But, if this is the standard for Slicer type annotation convention then I will change it.

File: Modules/Scripted/DICOMLib/DICOMProcesses.py
Comment (tbirdso) [reply Modules/Scripted/DICOMLib/DICOMProcesses.py]: Updated

File: Modules/Scripted/DICOMLib/DICOMUtils.py
Comment (tbirdso) [reply Modules/Scripted/DICOMLib/DICOMUtils.py]: Updated. Note that other functions in this file are broken up with this divider though.

**Comment by tbirdso** [comment]
From offline conversation with @jcfr I will work to add instructions for testing `slicer.util` and DICOMweb contributions. Unfortunately test automation is beyond the scope of this effort for now.

These changes include refactoring of `DICOMSender` DIMSE code paths. It would be good to test that functionality to ensure that refactoring did not introduce DIMSE behavioral changes. @pieper , do you maybe have suggestions on how to test DIMSE stow functionality?

**Comment by lassoan** [comment]
Probably the easiest way to test DICOM store is to run DCMTK's storescp (C-STORE service class provider).

**Comment by pieper** [comment]
>  @pieper , do you maybe have suggestions on how to test DIMSE stow functionality?

You can have a look at how query / retrieve are tested in CTK.  We run the dcmtk executables as @lassoan suggested and configure it all via cmake:

https://github.com/commontk/CTK/blob/master/Applications/Testing/Cpp/CMakeLists.txt

https://github.com/commontk/CTK/blob/master/Applications/Testing/Cpp/ctkDICOMApplicationTest1.cpp

File: Base/Python/slicer/util.py
Comment (jcfr) [reply Base/Python/slicer/util.py:3802]: > I'm not sure I understand the suggested change to remove the None return type annotation?

You are correct. This was on oversight on my side.

The one we should omit is the annotation for `Self`. See https://docs.astral.sh/ruff/rules/missing-type-self/

**Comment by tbirdso** [comment]
Thanks @lassoan @pieper for pointing me in the right direction. I've tested refactored DICOM networking code paths with the following approaches:
- DIMSE: ran `storescp` on my local system at @lassoan 's suggestion, exported sample data from the Slicer DICOM database to `localhost` and confirmed files populate in the `storescp` working directory
- DICOMweb: I set up a local [Orthanc](https://www.orthanc-server.com/download-windows.php) instance with its DICOMweb plugin and tested both with and without authentication. Confirmed that export with `DICOMSender` fails with missing/incorrect global authentication information and succeeds with correct auth information.

Unfortunately I do not have resources to pursue automated testing at this time. However, I've added a note to DICOM developer documentation to help reproduce similar manual testing in the future.

I've added a few extra documentation updates and QOL improvements reflective of that testing. Please take a moment to (re-)review at your convenience. cc @jcfr

File: Modules/Scripted/DICOMLib/DICOMProcesses.py
Comment (jcfr) [inline Modules/Scripted/DICOMLib/DICOMProcesses.py:509]:
```suggestion
```

The original logic introduced in https://github.com/Slicer/Slicer/commit/b89b3c8a4b2a209ba3654377ee3050da62a14877 and updated in https://github.com/Slicer/Slicer/commit/c25b4e6212849b14f8d2c69324b4ee88b6abd558.

It was intended to check if the version of `dicomweb_client` provided  by Slicer needed to be updated.

Since then, the version has been updated[^1] and this logic can simply be removed.

[^1]: https://github.com/Slicer/Slicer/commits/main/SuperBuild/External_python-dicom-requirements.cmake

File: Base/Python/slicer/util.py
Comment (jcfr) [inline Base/Python/slicer/util.py:3860]:
```suggestion
```

I suggest we revisit the generalization of in a another pull request and simply remove the check currently added in `DICOMProcesses`

Instead of `validate_module_version`, having a method called `is_python_module_installed` would be more explicit and focused on doing on task: _Checking that a specific module (and optinally specified version) is installed_

Strictly speaking[^1], raising `ModuleNotFoundError` is not _valid_ in the sense that a module could be installed without having the correct version. I am not sure redefining the meaning of this specific exception in he context of this function is sensible.

Installation, confirmation[^2], progress reporting and restart[^3] should be handled outside of this function.

### Example

An idea would be to support the following:

```
moduleName="dicomweb_client"
versionSpecifier = ">=0.51"

if not slicer.util.is_python_module_installed(
    moduleName, version_specifier=versionSpecifier
):
    if slicer.util.confirmOkCancelDisplay("This current operation requires '{module_name}' Python package. Click OK to install it now."):
        slicer.util.pip_install(f"{moduleName}{versionSpecifier}")
```


[^1]: https://docs.python.org/3/library/exceptions.html#ModuleNotFoundError

[^2]: Handle using the approach suggested in at https://slicer.readthedocs.io/en/latest/developer_guide/script_repository.html#install-a-python-package

[^3]: Restart should also be handled with a confirmation dialog.

File: Modules/Scripted/DICOMLib/DICOMProcesses.py
Comment (tbirdso) [reply Modules/Scripted/DICOMLib/DICOMProcesses.py:509]: Thanks for the background. I agree that this check should be unnecessary in the current Slicer build. My goal in this PR has been to introduce authentication paths for DICOMweb integration and otherwise improve code structure with minimal possible changes to functionality/interface. But, if it is more convenient I will remove the unneeded check here.

File: Base/Python/slicer/util.py
Comment (tbirdso) [reply Base/Python/slicer/util.py:3860]: > I suggest we revisit the generalization of in a another pull request

That's fine. I won't have bandwidth to drive refinement in that separate PR, but if you think this is a worthwhile utility to pursue I can open a draft PR for someone else to take across the finish line.

> and simply remove the check currently added

Will do.

Comment (tbirdso) [reply Base/Python/slicer/util.py]: Tracked in https://github.com/Slicer/Slicer/pull/7268 and removed from this PR.

File: Modules/Scripted/DICOMLib/DICOMProcesses.py
Comment (tbirdso) [reply Modules/Scripted/DICOMLib/DICOMProcesses.py]: The `dicomweb_client` version check is now removed from `DICOMSender` in this PR.

**Review by sjh26 — APPROVED** [review]
lgtm

**Comment by tbirdso** [comment]
On review of `CONTRIBUTING.md` guidelines I will wait for an additional approval before merge.

> Consensus is reached. This usually means that at least two reviewers approved the changes

**Comment by tbirdso** [comment]
@jcfr Ping for review

**Review by jcfr — APPROVED** [review]
Thanks for revisiting the patches :pray: :100: :tada:

**Pull request merged**
